### DISCUSSION TITLE
How to Constrain Subsets of an Assignment Matrix Idiomatically?

### DISCUSSION THREAD

**Discussion by TBulka81** [dirkschumacher/ompr#449] [Q&A] [answered]
Hi everyone,

I want to model a specific assignment problem, but unfortunately I'm not too fluent in R, so I'm struggling to find the most idiomatic way to do so.

Say, I add an assignment matrix `x` of binary decision variables to my model like this:

```
n <- 10
m <- 12

model <- MIPModel() %>%
  add_variable(x[i,j], i = 1:n, j = 1:m, type = "binary")
```

What I want to do now, is to add a set of constraints, which have the following structure (pseudocode):

`sum(x[1,2] + x[3,1] + x[2,8] + ...) <= someNumber`

So basically these constraints shall enforce, that the sum of some more or less arbitrary set of variables in `x` is smaller than some limit.

My question now is: What would be a good data structure in R to bring this information (for each constraint the tuples which carry the constrained indices and the limit to enforce) into model construction with ompr? And how could these constraints be added to the model automatically by looping over this input data?

I guess there is a way to do so seamlessly, without the requirement to add each of the constraints manually, but I'm a little bit stuck in finding the best way to do so in R.

Thanks very much for your help in advance!

Regards,

Thomas

**Comment by dirkschumacher** [comment, marked as answer]
Sorry for the late reply 🙈

`ompr` uses the package `listcomp` to build constraints and summations. `listcomp` is a package that helps you build lists of things. If you want to sum over arbitrary tuples you can use a named list for that.

```r
vars <- list(
  i = c(1,3,2),
  j = c(2, 1, 8)
)
model <- add_constraint(
  model, 
  sum_over(x[i,j], list(i = vars$i, j = vars$j)) <= 42
)

listcomp::gen_list(c(i,j), list(i = vars$i, j = vars$j))
#[[1]]
#[1] 1 2
#[[2]]
#[1] 3 1
#[[3]]
#[1] 2 8
```

The `list(i = vars$i, j = vars$j)` means the `sum_over` will iterate over `i` and `j` in parallel. In contrast to `sum_over(x[i,j], i = vars$i, j = vars$j) <= 42` which iterates over all combinations of `i` and `j`.

You can also combine that by the way and also add filters to form very complex sets. E.g.
```r
sum_over(x[i,j,k], list(i = vars$i, j = vars$j), k = 1:10, i < k)
```
